concreate/builders/osbs.py

```diff
@@ -0,0 +1,206 @@
+import os
+import shutil
+import subprocess
+
+from concreate.builder import Builder
+from concreate.errors import ConcreateError
+
+logger = logging.getLogger('concreate')
+
+
+class OSBSBuilder(Builder):
+    """Class representing OSBS builder."""
+
+    def prepare(self, descriptor):
+        """Prepares dist-git repository for OSBS build."""
+
+        repository_key = descriptor.get('osbs', {}).get('repository', {})
+        repository = repository_key.get('name')
+        branch = repository_key.get('branch')
+
+        if not (repository and branch):
+            raise ConcreateError("OSBS builder needs repostiory and branch provided!")
+
+        self.dist_git_dir = os.path.join(os.path.expanduser('~/.concreate.d'),
+                                         'osbs',
+                                         repository)
+        if not os.path.exists(os.path.dirname(self.dist_git_dir)):
+            os.makedirs(os.path.dirname(self.dist_git_dir))
+
+        self.dist_git = Git(self.dist_git_dir,
+                            self.target,
+                            repository,
+                            branch)
+
+        self.dist_git.prepare()
+        self.dist_git.clean()
+
+        self.update_osbs_image_source()
+
+        artifacts = [a['name'] for a in descriptor.get('artifacts', [])]
+        self.update_lookaside_cache(artifacts)
+
+    def update_osbs_image_source(self):
+        for obj in ["repos", "modules", "Dockerfile"]:
+            if os.path.exists(obj):
+                shutil.copytree(os.path.join(self.target,
+                                             'image',
+                                             obj),
+                                os.path.join(self.dist_git_dir,
+                                             obj))
+
+    def update_lookaside_cache(self, artifacts):
+        if not artifacts:
+            return
+        for artifact in artifacts:
+            shutil.copy(os.path.join(self.target,
+                                     'image',
+                                     artifact),
+                        os.path.join(self.dist_git_dir,
+                                     artifact))
+        logger.info("Updating lookaside cache...")
+        cmd = ["rhpkg", "new-sources"] + artifacts
+        logger.debug("Executing '%s'" % cmd)
+        with Chdir(self.dist_git_dir):
+            subprocess.check_output(cmd)
+        logger.info("Update finished.")
+
+    def build(self, build_args):
+        build_cmd = ["rhpkg", "container-build"]
+        if not build_args.build_osbs_release:
+            build_cmd.append("--scratch")
+
+        with Chdir(self.dist_git_dir):
+            self.dist_git.add()
+
+            if self.dist_git.stage_modified():
+                self.dist_git.commit()
+                self.dist_git.push()
+            else:
+                logger.info("No changes made to the code, committing skipped")
+
+            logger.info("Executing container build on OSBS...")
```

Comment:
We should NOT execute a build in case we don't push changes to dist-git repo. In other words, when we answer "N" when asked about pushing to repo - we cannot kick a build and builder should immediately exit.